### PR TITLE
[DashboardLayout] Add hideNavigation prop

### LINKED ISSUE #4229: Allow hiding the navigation bar and toggle button
### Summary

Allow hiding the navigation bar and toggle button in case we just want to show the header

### Examples

_No response_

### Motivation

In my app I want to use toolpad to handle navigation and header after a user logs in. But I have a case where user needs to wait before they are verified to see other pages, and on that page I just need the header.

I guess I can use the AppBar component from MUI but that would require me to handle `account` flow separately. 

**Search keywords**: hide navigation

### PR DESCRIPTION
**Pull request by vikasgurjar**
Closes #4229

- [x] I've read and followed the [contributing guide](https://github.com/mui/toolpad/blob/master/CONTRIBUTING.md#sending-a-pull-request) on how to create great pull requests.
- [x] I've updated the relevant documentation for any new or updated feature.
- [x] I've linked relevant GitHub issue with "Closes #<issue id>".
- [x] I've added a visual demonstration in the form of a screenshot or video.

image

https://deploy-preview-4231--mui-toolpad-docs.netlify.app/toolpad/core/react-dashboard-layout

### CODE REVIEW

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 1]
```diff
@@ -594,44 +604,46 @@ function DashboardLayout(props: DashboardLayoutProps) {
...
-      >
-        {getDrawerContent(isMobileMini, 'Tablet')}
-      </Drawer>
-      <Drawer
-        variant="permanent"
-        sx={{ display: { xs: 'none', md: 'block' }, ...getDrawerSharedSx(isDesktopMini) }}
-      >
-        {getDrawerContent(isDesktopMini, 'Desktop')}
-      </Drawer>
+      {!hideNavigation && <React.Fragment>
```
Comment (apedroferreira) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:607]: `React.Fragment` not needed?

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 2]
```diff
@@ -360,6 +360,11 @@ export interface DashboardLayoutProps {
     toolbarActions?: {};
     toolbarAccount?: AccountProps;
   };
+  /**
+  * If the navigation bar and button should to be hidden
```
Comment (apedroferreira) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:364]: Can we rephrase it to `Whether the navigation bar and menu icon should be hidden`?

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 3]
```diff
@@ -460,6 +465,11 @@ function DashboardLayout(props: DashboardLayoutProps) {
 
   const getMenuIcon = React.useCallback(
     (isExpanded: boolean) => {
+
+      if (hideNavigation) {
```
Comment (apedroferreira) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:469]: There's also a `marginRight` on the boxes around these icons that we should not have if they're not shown.
So instead on line 566 we can probably have something like:

```tsx
{!hideNavigation ? 
         <Box
            sx={{
              mr: { sm: disableCollapsibleSidebar ? 0 : 1 },
              display: { md: 'none' },
            }}
          >
            {getMenuIcon(isMobileNavigationExpanded)}
          </Box>
          <Box
            sx={{
              display: { xs: 'none', md: disableCollapsibleSidebar ? 'none' : 'block' },
              mr: disableCollapsibleSidebar ? 0 : 1,
            }}
          >
            {getMenuIcon(isDesktopNavigationExpanded)}
          </Box> : null}
```

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx (hunk 1 above)
Comment (apedroferreira) [reply packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:607]: Also can we write it like `{!hideNavigation ? ... : null}`? Just because we've been following that convention in the project.

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.test.tsx [hunk 4]
```diff
@@ -340,4 +340,35 @@ describe('DashboardLayout', () => {
...
+    render(
+      <AppProvider navigation={NAVIGATION}>
+        <DashboardLayout hideNavigation>Hello world</DashboardLayout>
+      </AppProvider>,
+    );
+  
+    const desktopNavigation = screen.queryByRole('navigation', { name: 'Desktop' });
+    const navigationToggle = screen.queryByLabelText('Collapse menu');
+  
+    // Expect that the navigation and toggle button is not rendered
```
Comment (apedroferreira) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.test.tsx:367]:
```suggestion
    // Expect that the navigation and toggle button are not rendered
```

File: docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md [hunk 5]
```diff
@@ -109,6 +109,12 @@ The layout sidebar is collapsible to a mini-drawer (with icons only) in desktop
 
 {{"demo": "DashboardLayoutNoMiniSidebar.js", "height": 400, "iframe": true}}
 
+### Hiding the sidebar
+
+The layout sidebar can be hidden if needed using the `hideNavigation` prop.
```
Comment (apedroferreira) [inline docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:114]:
```suggestion
The layout sidebar can be hidden if needed with the `hideNavigation` prop.
```

**Review by apedroferreira — CHANGES_REQUESTED** [review]
Looks perfect, thank you! And that's exactly what I meant with the demo, good job finding it! :)
Just a few really minor changes suggested.

Also for the docs changes you can run:
`pnpm proptypes` to generate proptypes the be read by the API documentation
`pnpm docs:build:api` to build the API docs
`pnpm docs:typescript:formatted` to format all the docs files based on the .js file

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutSidebarHidden.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import * as React from 'react';
+
+import { AppProvider } from '@toolpad/core/AppProvider';
+import Box from '@mui/material/Box';
+import { DashboardLayout } from '@toolpad/core/DashboardLayout';
+import PropTypes from 'prop-types';
+import Typography from '@mui/material/Typography';
+import { createTheme } from '@mui/material/styles';
+
+const NAVIGATION = [];
```
Comment (apedroferreira) [inline docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutSidebarHidden.js:10]: We can just not pass this as it defaults to []

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutSidebarHidden.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
+import * as React from 'react';
+
```
Comment (apedroferreira) [inline docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutSidebarHidden.js:2]: empty line

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutSidebarHidden.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+DemoPageContent.propTypes = {
+  pathname: PropTypes.string.isRequired,
+};
+
+function DashboardLayoutSidebarHidden(props) {
+  const { window } = props;
+
+  const [pathname, setPathname] = React.useState('/dashboard');
+
+  const router = React.useMemo(() => {
```
Comment (apedroferreira) [inline docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutSidebarHidden.js:53]: No need to pass a router in this example too I think.

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutSidebarHidden.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+      searchParams: new URLSearchParams(),
+      navigate: (path) => setPathname(String(path)),
+    };
+  }, [pathname]);
+
+  // Remove this const when copying and pasting into your project.
+  const demoWindow = window !== undefined ? window() : undefined;
+
+  return (
+    // preview-start
```
Comment (apedroferreira) [inline docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutSidebarHidden.js:65]: You can remove `// preview-start` and `// preview-end` here and the code preview will show only the DashboardLayout which is the most important information.

**Comment by vikasgurjar** [comment]
@apedroferreira I will need your help with the `test_static` build issue. Due to some conf mismatch I'm not able to fix the above issue. Can you guide on how to fix it?

**Comment by apedroferreira** [comment]
> @apedroferreira I will need your help with the `test_static` build issue. Due to some conf mismatch I'm not able to fix the above issue. Can you guide on how to fix it?

Can you run `pnpm docs:typescript:formatted`? That should do it.
Everything looks great btw, thanks for the changes!

**Comment by vikasgurjar** [comment]
> > @apedroferreira I will need your help with the `test_static` build issue. Due to some conf mismatch I'm not able to fix the above issue. Can you guide on how to fix it?
> 
> Can you run `pnpm docs:typescript:formatted` from the project root? That should do it. Everything looks great btw, thanks for the changes!

I have done that multiple times now. I'm running it on windows, not really what's wrong here, can it be new line char issue or formatter issue?

**Comment by apedroferreira** [comment]
> > > @apedroferreira I will need your help with the `test_static` build issue. Due to some conf mismatch I'm not able to fix the above issue. Can you guide on how to fix it?
> > 
> > 
> > Can you run `pnpm docs:typescript:formatted` from the project root? That should do it. Everything looks great btw, thanks for the changes!
> 
> I have done that multiple times now. I'm running it on windows, not really what's wrong here, can it be new line char issue or formatter issue?

I see, when I run `pnpm docs:typescript:formatted` (`docs/scripts/formattedTSDemos` script) on my Macbook there's a difference in the `DashboardLayoutSidebarHidden.tsx.preview` file compared to your current PR... Might be an OS thing, maybe line endings.

@mui/docs-infra any knowledge on this issue? Is it an issue with the script? We're generally using the same script as https://github.com/mui/material-ui/blob/master/docs/scripts/formattedTSDemos.js.

**Comment by Janpot** [comment]
The line endings don't match. Maybe you hit "save" on the preview and the auto-formatter kicked in?

**Comment by vikasgurjar** [comment]
> The line endings don't match. Maybe you hit "save" on the preview and the auto-formatter kicked in?

There was no need to hit the save. Seems like some other issue. Even your commit didn't help here

**Comment by Janpot** [comment]
It did fix it, CI just found another issue, still had to run `pnpm prettier`.

**Comment by vikasgurjar** [comment]
> It did fix it, CI just found another issue, still had to run `pnpm prettier`.

Great, Thanks! I was setting up WSL in my windows machine, to try fixing this

**Comment by apedroferreira** [comment]
Ok, let me know if you ever find what the issue was! Just added some small demo changes and will merge :)

**Review by apedroferreira — APPROVED** [review]
